File: src/main/java/me/jollyperson/featurebot/objects/BotUser.java
package me.jollyperson.featurebot.objects;

public class BotUser {

    private long id;
    private BotEconomyUser eco;
    private String permission;

    public long getId() {
        return id;
    }

    public BotEconomyUser getEco() {
        return eco;
    }

    public BotUser(long id, BotEconomyUser eco, CommandPermission perm) {
        this.id = id;
        this.eco = eco;
        this.permission = perm.toString();
    }

    public CommandPermission getPerm() {
        return CommandPermission.valueOf(permission);
    }
}

File: src/main/kotlin/me/jollyperson/featurebot/objects/BotUser.kt
package me.jollyperson.featurebot.objects

import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.Permission

data class BotUser(val id: Long, val eco: BotEconomyUser, val groups: MutableCollection<BotRole>,val userPerms: MutableList<PermissionFlag>) {
    /**
     * Returns the group(s) of this player.
     *
     * @return      the groups of this user
     * @see Group
     */

    fun addPermission(permissionFlag: PermissionFlag) {
        userPerms.add(permissionFlag)
    }

    fun removePermission(permissionFlag: PermissionFlag?) {
        userPerms.remove(permissionFlag)
    }

    fun hasPermission(flag: PermissionFlag?, jda: JDA, guildID: Long, allowRoleOverrides: Boolean): Boolean {
        return if (userPerms.contains(flag)) {
            if (allowRoleOverrides) {
                jda.getGuildById(guildID)!!.getMemberById(id)!!.hasPermission(Permission.ADMINISTRATOR)
            } else true
        } else false
    }

    fun hasPermission(flag: PermissionFlag?): Boolean {
        val isPerm = userPerms.contains(flag)
        if(!isPerm){
            groups.forEach { _ -> return hasPermission(flag) }
        }
        return isPerm
    }

    init {
        groups
    }
}